src/ngraph/pass/nop_elimination.cpp

```diff
@@ -134,62 +134,263 @@ static bool eliminate_reshape_v1(const std::shared_ptr<Node>& node)
+        if (shape_ps[i].is_dynamic())
+        {
+            target_shape.emplace_back(-1);
+        }
+        else
+        {
+            target_shape.emplace_back(shape_ps[i]);
+        }
+    }
+
+    shared_ptr<Node> reshape;
+    auto input = node->input_value(0).get_node_shared_ptr();
+    auto pat =
+        op::Constant::create<int64_t>(element::i64, Shape{target_shape.size()}, target_shape);
+
+    if (is_type<opset3::Reshape>(input) || is_type<opset3::Squeeze>(input) ||
+        is_type<opset3::Unsqueeze>(input))
+    {
+        reshape = make_shared<opset3::Reshape>(input->input_value(0), pat, false);
+    }
+    else
+    {
+        reshape = make_shared<opset3::Reshape>(node->input_value(0), pat, false);
+    }
+
+    // skip if reshape is nop
+    if (reshape->get_input_partial_shape(0).same_scheme(shape_ps))
+    {
+        return replace_output_update_name(node->output(0), reshape->input_value(0));
+    }
+    else
+    {
+        return replace_output_update_name(node->output(0), reshape->output(0));
+    }
+    return false;
+}
+
+static std::vector<int64_t> get_unsqueeze_axes(const PartialShape& data_shape,
+                                               const PartialShape& out_shape)
+{
+    std::vector<int64_t> axes;
+    size_t i = 0;
+    for (auto o = 0; o < out_shape.rank().get_length(); o++)
+    {
+        if (i < data_shape.rank().get_length() && data_shape[i].same_scheme(out_shape[o]))
+        {
+            i += 1;
+            continue;
+        }
+        if (out_shape[o].is_static() && out_shape[o] == 1)
+        {
+            axes.push_back(o);
+        }
+    }
+    return axes;
+}
+
+static std::vector<int64_t> get_squeeze_axes(const PartialShape& data_shape,
+                                             const PartialShape& out_shape)
+{
+    std::vector<int64_t> axes;
+    size_t out_i = 0;
+    for (auto i = 0; i < data_shape.rank().get_length(); i++)
+    {
+        if (out_i < out_shape.rank().get_length() && data_shape[i].same_scheme(out_shape[out_i]))
+        {
+            out_i += 1;
+            continue;
+        }
+        if (data_shape[i].is_static() && data_shape[i] == 1)
+        {
+            axes.push_back(i);
+        }
+    }
+    return axes;
+}
+
 static bool eliminate_unsqueeze(const std::shared_ptr<Node>& node)
 {
+    auto out_shape = node->get_output_partial_shape(0);
+    // try to replace all squeeze/unsqueeze with reshape
+    if (out_shape.rank().get_length() != 0 && count_unknown_dims(out_shape) < 2)
+    {
+        return replace_squeeze_unsqueeze(node);
+    }
+
     auto unsqueeze = as_type_ptr<opset3::Unsqueeze>(node);
     auto input = unsqueeze->input_value(0).get_node_shared_ptr();
+    auto data_shape = input->input_value(0).get_partial_shape();
     auto squeeze = as_type_ptr<opset3::Squeeze>(input);
+    auto replace_unsqueeze_only = [&](const vector<int64_t>& axes) {
+        auto axes_const = op::Constant::create<int64_t>(element::i64, Shape{axes.size()}, axes);
```

Comment:
I am not sure how important this but should we use v3::Constant everywhere instead of op::Constant?

@diyessi @GlebKazantaev could you comment?